Cython/Includes/libcpp/memory.pxd

```diff
@@ -59,6 +59,7 @@ cdef extern from "<memory>" namespace "std" nogil:
         shared_ptr(shared_ptr[T]&, T*)
         shared_ptr(unique_ptr[T]&)
         #shared_ptr(weak_ptr[T]&) # Not Supported
+        T& operator=[Y](const shared_ptr[Y]& ptr)
```

Comment:
Shouldn't this return a `shared_ptr[T]&` ?

And arguably there's a matching constructor too? (although a little harder to use from Cython since most things end up default-constructed)

Reply:
Thanks for the review!

> Shouldn't this return a `shared_ptr[T]&`?

Hm, again, I'm a Cython newb so I'm not sure -- certainly what you're saying looks more logical and closer to [what's in the reference](https://cplusplus.com/reference/memory/shared_ptr/operator=/). But I've tried both and they generate the same cpp in the case of the `test_assignment_to_base_class` example.

> And arguably there's a matching constructor too?

Do you mean `shared_ptr[Y](shared_ptr[Y]&)`? As the SO poster [remarks](https://stackoverflow.com/a/67631211/5085211), that doesn't quite work.

```
        shared_ptr[Y](shared_ptr[Y]&)
                    ^
------------------------------------------------------------
memory.pxd:64:21: Missing name in declaration.
```

Reply:
> Hm, again, I'm a Cython newb so I'm not sure -- certainly what you're saying looks more logical and closer to what's in the reference. But I've tried both and they generate the same cpp in the case of the test_assignment_to_base_class example.

It should only matter in for multiple assignments (or similar). So for `test_assignment_to_base_class` it makes no difference, but for `second_copy = first_copy = original` it might make a difference (I think... I haven't looked into the exact code that Cython generates here).

> Do you mean `shared_ptr[Y](shared_ptr[Y]&)`? As the SO poster remarks, that doesn't quite work.

I'd think it should be `shared_ptr(shared_ptr[Y]&)` (note the first square brackets gone). If that doesn't work then it's fine - don't spend more time on it though. The point the SO poster is making is that Cython generally generates c++ code like:

```
// start of function
shared_ptr<A> a;
// function body
a = something;
```

and so `operator=` is generally much more useful.

Reply:
>  for `second_copy = first_copy = original` it might make a difference

I just tried

```python
    cdef shared_ptr[C] derived = shared_ptr[C](new C())
    cdef shared_ptr[A] base1
    cdef shared_ptr[A] base2
    base1 = base2 = derived
```

Here, the generated code will once again be the same for both return types. In any case, I've changed it to `shared_ptr[T]&`.

> If that doesn't work then it's fine

Did try that one as well; doesn't quite work:

```
        shared_ptr(shared_ptr[Y]&)
                              ^
memory.pxd:64:31: unknown type in template argument
```

> don't spend more time on it though

I don't mind; this is quite helpful for learning what's going on!